### ISSUE TITLE
Gunicorn and gevent worker support

### ISSUE THREAD

**Issue by itsderek23** [scoutapp/scout_apm_python#28]
We haven't tested the agent with [async workers via gevent](http://docs.gunicorn.org/en/stable/design.html#async-workers) yet. 

We should.

**Comment by svvitale** [comment]
So my use case is slightly different with gunicorn + uvicorn (which is an async worker using uvloop under the hood).  I'm guessing that the issue may be with Async workers in general, so I thought I would post these details here.  If this belongs in its own issue, please let me know.

The setup for scout-apm appears to work correctly, and I see debug messages that look reasonable:

[scout_apm_debug.log](https://github.com/scoutapp/scout_apm_python/files/3267875/scout_apm_debug.log)

I never see any request data show up in the Scout dashboard though.

Python 3.6
Django 2.0.3
Channels 2.1.5
gunicorn 19.9.0
uvicorn 0.4.6

This is all running on Heroku and uses the following Procfile line for starting the web process:

`gunicorn server.asgi -w $WEB_WORKERS -k uvicorn.workers.UvicornWorker --log-file -`

asgi.py:
```
"""
ASGI entrypoint. Configures Django and then runs the application
defined in the ASGI_APPLICATION setting.
"""
import os
import django
from channels.routing import get_default_application

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "server.settings")
django.setup()
application = get_default_application()
```

routing.py:
```
from django.conf.urls import url
from channels.routing import URLRouter
from channels.http import AsgiHandler
from channels.auth import AuthMiddlewareStack
import django_eventstream
from channels.routing import ProtocolTypeRouter

urlpatterns = [
    url(r"^channels/", AuthMiddlewareStack(URLRouter(django_eventstream.routing.urlpatterns))),
    url(r"", AsgiHandler),
]

application = ProtocolTypeRouter({"http": URLRouter(urlpatterns)})
```

Let me know if I can provide any further detail.  I've dug in as far as to verify that the middleware is being executed, but I'm not sure where else to look from here.

Thanks!

**Comment by adamchainz** [comment]
Thanks for the detailed report @svvitale . I'll chat with the other scout members and prioritize supporting ASGI.

**Comment by svvitale** [comment]
Just wanted to check-in and see if there's been any priority allocated to ASGI support?  I'm happy to be the guinea pig here.

**Comment by adamchainz** [comment]
Hi @svvitale 

I have an ASGI (3) implementation in progress in #285 . I should be able to make it work with ASGI 2 (which is Channels < 2.2.0) with a little extra effort.

After the generic ASGI implementation, I'll research how best to fit into Django + Channels.

Thanks,

Adam

**Comment by svvitale** [comment]
Awesome to hear, @adamchainz!  I can certainly move up to a newer version of Channels if that will help things along and give you a real world test case.  I took a quick look at the PR, but I'm not familiar enough with the Django internals to know how to plug it in.  Thanks for your work on this!

**Comment by adamchainz** [comment]
@svvitale I just released version 2.8.0 which has Starlette support - as a pure ASGI framework it was easier to work with. However I think it should also correctly track context within your uvicorn workers for your Django views, if you're using asgiref 3.1+.

For complete instrumentation of other Channels consumers, such as the websockets you're using, I have opened issue #372 .

I'll keep this issue open for gevent support, which is a separate async method to ASGI/asyncio.

**Comment by adamchainz** [comment]
I just gave gunicorn with gevent and eventlet workers a quick test using my [plain django test app](https://github.com/adamchainz/scout-test-apps/tree/master/django). I hammered it with `parallel http` to force parallel requests. The tracing seemed to work just fine. Would be nice to add some tests to the suite but they would need a new tox environment that patches gevent / eventlet in place for the whole test run.
